Add a new /digress slash command.

abhijeetbodas2001 opened the pull request:
#14531
This PR adds a new digress slash command which will send two breadcrumb messages to old and new threads.
CZO topic: https://chat.zulip.org/#narrow/stream/101-design/topic/.2Fdigress
Screenshots
<details><summary>Successful digress (GIF)</summary>
<img src="https://user-images.githubusercontent.com/55339528/102993248-cdedbd80-4542-11eb-8c81-6d2e419163b3.gif" />

</details>

<details><summary>Successful digress (Feedback widget)</summary>
<img src="https://user-images.githubusercontent.com/55339528/101726140-f94ec200-3ad7-11eb-8151-76d5304364f0.png" />
</details>

<details><summary>Successful digress (Notification messages)</summary>
<img src="https://user-images.githubusercontent.com/55339528/101726599-d83aa100-3ad8-11eb-94d0-42554d07abed.png" />
</details>

<details><summary>Errors (above composebox)</summary>
<img src="https://user-images.githubusercontent.com/55339528/102992772-df829580-4541-11eb-8a90-8fc5614a6e45.png" />

<img src="https://user-images.githubusercontent.com/55339528/102992694-b235e780-4541-11eb-9ec5-cd6c271b1100.png" />


</details>

showell commented on zerver/lib/actions.py:
```diff
@@ -4801,8 +4803,14 @@ def user_info(um: UserMessage) -> Dict[str, Any]:
+        old_thread_notification_string = _("This topic was moved by {user} to {new_location}")
+        new_thread_notification_string = _("This topic was moved here from {old_location} by {user}")
+        notify_topic_moved_streams(user_profile,
+                                   stream_being_edited, orig_topic_name,
```
A minor formatting things--we are eventually moving toward just having a single argument per line for function calls using 3+ arguments.  It takes a little more vertical space, but it's more diff-friendly, and it makes it easier for most people to visually scan the arguments.

abhijeetbodas2001 replied:
Sure, I'll do that!

showell commented:
@amanagr This might be a quick review for you.  This seems like a reasonable prep refactor, and it only affects one function call.

abhijeetbodas2001 commented:
I addressed the formatting point, and also renamed `notify_topic_moved_streams` to a less specific `send_topic_change_breadcrumbs`.

abhijeetbodas2001 commented:
@amanagr @showell This is ready for review!
I would like to get you opinions on what the behavior should be when attempting to digress from/to an unsubscribed stream.
Currently, I have just shown a `compose_error` for these cases, but should we allow this?

abhijeetbodas2001 commented:
@amanagr @showell This PR is ready. Please review it!

amanagr commented on zerver/lib/actions.py:
```diff
@@ -4382,11 +4382,15 @@ class MessageUpdateUserInfoResult(TypedDict):
     message_user_ids: Set[int]
     mention_user_ids: Set[int]
 
-def notify_topic_moved_streams(user_profile: UserProfile,
```
Split this into two commits -> Rename of function and allowing using of custom msg strings. Maybe `send_message_moved_breadcrumbs` is a better name for this function since I expect we will use this function when moving both topics and streams in the future.

amanagr commented on zerver/lib/zcommand.py:
```diff
@@ -54,4 +58,47 @@ def change_mode_setting(command: str, switch_command: str,
+        if not data or not all(key in data for key in data_keys):
+            raise JsonableError(_('Invalid data.'))
+
+        (old_stream, ignored_old_stream_sub) = access_stream_by_name(
```
Use `_` for ignored variables. That is the pythonic way to do it.

abhijeetbodas2001 replied:
I used this style after seeing it used in other parts of the codebase ([example](https://github.com/zulip/zulip/blob/master/zerver/views/users.py#L372)). Since someone might do the same in the future, would it be worth/possible changing these instances and/or enforcing this in linters?

amanagr replied:
Ohh! I am at fault here. We cannot use `_` here since `_` is used for text that has to be translated. That is why Steve is using those igonored_ variables too.

amanagr commented on zerver/lib/zcommand.py:
```diff
@@ -54,4 +58,47 @@ def change_mode_setting(command: str, switch_command: str,
+                msg=_('Old topic #**{}>{}** does not exist.').format(data["old_stream"], data["old_topic"])
+            )
+
+        if data["old_stream"] == data["new_stream"]:
+            if data["old_topic"] == data["new_topic"]:
```
Why not do this in the same line? Also, since checking this doesn't require access to stream, we should move it next to 'Invalid data` check.

abhijeetbodas2001 replied:
I was worried about line length, but this didn't turn out to be any longer than some of the other lines. Updated this, and moved it up, along with it's test.

amanagr commented on zerver/tests/test_zcommand.py:
```diff
@@ -81,3 +83,67 @@ def test_fixed_zcommand(self) -> None:
+        )
+        result = self.client_post("/json/zcommand", payload)
+        self.assert_json_success(result)
+        self.assertEqual('', result.json()['msg'])
```
You should verify what was the message that was sent in the both the threads here.

amanagr commented on static/js/composebox_typeahead.js:
```diff
@@ -365,6 +365,10 @@ exports.slash_commands = [
         name: "day",
     },
+    {
+        text: i18n.t("/digress (Shift conversation to another topic)"),
```
We should inform user about the format in which we expect the new stream and topic is to be supplied. We should discuss this in czo.

abhijeetbodas2001 replied:
Updated this typeahed to be `/digress #stream name>topic name (Shift conversation to another topic)`

amanagr commented on zerver/lib/zcommand.py:
```diff
@@ -54,4 +58,47 @@ def change_mode_setting(command: str, switch_command: str,
+            "new_topic",
+        ]
+        if not data or not all(key in data for key in data_keys):
+            raise JsonableError(_('Invalid data.'))
```
Also, inform user what is the correct use of /digress command. Maybe with an example.

amanagr reviewed (commented):
@abhijeetbodas2001 Thanks for working on this! Left some comments.

abhijeetbodas2001 commented:
@amanagr Thanks for reviewing! I have updated the PR.

amanagr commented:
Looks fine to me, other than wrongly spelled `weird` in 3rd commit. We could check in the fronted itself if the params to /digress are valid so that user gets a snappy response, but let's wait for others to review.

abhijeetbodas2001 commented:
> Looks fine to me, other than wrongly spelled weird in 3rd commit.

Oops! Fixed it!

> We could check in the fronted itself if the params to /digress are valid so that user gets a snappy response

Yes, I think both the checks can be done in the frontend. And then we could return a proper `json_error` in the backend check instead of sending the error message in the `json_success` right? Shall I go ahead and do that?

amanagr commented:
On second though, we should leave it for server since we need to think about multi-client support here. We cannot expect every client to implement their version of a similar check.

amanagr commented:
@abhijeetbodas2001 I will be working on top of your PR for adding the /giphy zcommand. I need the additional data pipeline which you already have implemented in this PR.

amanagr commented on zerver/tests/test_zcommand.py:
```diff
@@ -81,3 +84,75 @@ def test_fixed_zcommand(self) -> None:
+    def test_digress_zcommand(self) -> None:
+        self.login('hamlet')
+        hamlet = self.example_user('hamlet')
+        hamlet.save()
```
I don't think we need hamlet.save here. We are not modifying any property of hamlet.

abhijeetbodas2001 replied:
Right, fixed.

abhijeetbodas2001 commented:
I rebased and resolved conflicts in the last push.
@showell @timabbott Please take a look.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5101,9 +5101,11 @@ def send_message_moved_breadcrumbs(
     new_stream: Stream,
     new_topic: Optional[str],
-    send_notification_to_old_thread: bool,
+    new_thread_notification_string: str,
```
I think a better interface would be to pass an `Optional[str]`, and treat `None` as the boolean value being false, rather than passing two semi-duplicate fields.

abhijeetbodas2001 replied:
Makes sense! Updated.

timabbott commented:
I merged the first commit as fc0488fdb1b83bffea4a300656d7bb7f5e6ab581 and posted a comment on the next one.  Thanks @abhijeetbodas2001!

abhijeetbodas2001 commented:
@timabbott I addressed your [comment above](https://github.com/zulip/zulip/pull/16589#discussion_r577261494) in the last push. This is ready for another review.

abhijeetbodas2001 commented:
The last push was a rebase after the conversion of `zcommand.js` to ES6

abhijeetbodas2001 commented on static/js/zcommand.js:
```diff
@@ -191,6 +241,42 @@ export function process(message_content) {
+            return true;
+        }
+
+        let param = content.slice(8);
+        param = param.trim();
+
+        // A valid param will now be of the form #**stream name>topic name**
+        const new_stream_name_start = param.indexOf("#**") + 3;
+        const new_topic_start = param.indexOf(">") + 1;
+        const new_topic_end = param.lastIndexOf("**");
```
Would it be better to use `marked(param)` (which will return something like `<p><a class="stream-topic" data-stream-id="14" href="/#narrow/stream/14-test/topic/topic">#test &gt; topic</a></p>`) and then use jQuery to fetch the stream-id and topic?

timabbott commented:
The first commit looks great, so I've merged it as a30ca8490d4aa88a9f9c004a921f2fda7b0a281b.

timabbott commented on zerver/lib/zcommand.py:
```diff
@@ -73,4 +77,52 @@ def change_mode_setting(
+            user_profile=user_profile,
+            recipient_id=old_stream.recipient_id,
+            topic_name=data["old_topic"],
+        )
```
This check deserves at least a comment, because it's not strictly correct -- a user could in theory be trying to digress in a public stream they're not subscribed to and thus don't have any UserMessage objects for.  

I think the fix would be to check if the user has access to full history for the stream via the standard helper, and in that case, do effectively `message_exists_for_topic` instead.

If it's easy which I think it is, best to just fix, but if it's hard, we can open a follow-up issue and leave a comment here.

timabbott commented on zerver/lib/zcommand.py:
```diff
@@ -73,4 +77,52 @@ def change_mode_setting(
+            return dict(
+                msg=_("Old topic #**{}>{}** does not exist.").format(
+                    data["old_stream"], data["old_topic"]
+                )
```
It's a little awkward that this is a `json_success` -- shouldn't we `raise JsonableError` here instead, since the user's command didn't work?

(I'm not super familiar with the slash command system's error handling conventions, so this comment could be wrong).

abhijeetbodas2001 replied:
Yes, makes sense. I changed all these to raise `JsonableError`. These will be shown above the compose box (in the second commit).

All slash commands currently send `success`, and the messages are latter shown in the feedback widget. I think that is because the current slash commands are simple enough that they can't really "fail".

timabbott commented on zerver/lib/zcommand.py:
```diff
@@ -73,4 +77,52 @@ def change_mode_setting(
+            )
+
+        old_thread_notification_string = _(
+            "{user} will be talking on the new topic: {new_location}"
```
Let's say "{user} digressed to the new topic {new_location}`.

abhijeetbodas2001 replied:
Changed this string.

timabbott commented on zerver/tests/test_zcommand.py:
```diff
@@ -80,3 +83,93 @@ def test_fixed_zcommand(self) -> None:
+        self.assertEqual(
+            messages[0].content,
+            f"@_**King Hamlet|{hamlet.id}** digressed this from old topic: #**general>old topic name**",
+        )
```
We'll want to expand the tests to have a case where one tries to digress from a message in a private stream without shared history that the user joined after the original message, which should fail.  And then maybe another for "private stream with shared history" where the user doesn't have a UserMessage -- cleanest would be for that to reuse the data setup and just toggle the stream to "shared history" in between attempts.

abhijeetbodas2001 replied:
I split this into multiple tests, because the `UserMessage` generated while testing the happy path for the shared history case would interfere while testing the failure condition of the private history case.

timabbott commented on zerver/lib/zcommand.py:
```diff
@@ -73,4 +77,52 @@ def change_mode_setting(
+            "new_topic",
+        ]
+        if not data or not all(key in data for key in data_keys):
+            raise JsonableError(_("Invalid data."))
```
Somewhere in here we should validate that `new_topic` is a valid topic (not too long, has been stripped), ideally reusing whatever helper we have from the `send_message` code path.  And make sure we test for that case.

(I think it'll currently fail in the `internal_send_message` below, but we possibly after sending one but not the other of the two messages, which is clearly no good).

abhijeetbodas2001 replied:
Done, I added a call to `truncate_topic` to clean the new topic.

timabbott commented:
OK, posted a batch of comments that should be sufficient for us to be able to merge the backend for this.

abhijeetbodas2001 commented:
@timabbott Thanks for the review! I've updated this PR, which was pretty much a rewrite of the first commit. I understand that the earlier implementation was far from thoughtful. I'll try to put more thought into permissions and the like henceforth, especially while working on anything which deals with message sending. Here are the changes since the last review-
- I've tried to handle all cases of streams being private/public and private streams having shared/protected history.
- I've also added `access_stream_for_message_send` checks.
- Split the one test into multiple parts. The tests here are getting pretty big, so I've added some comments to hopefully improve readability.

Apart from these changes, I have one more doubt. Currently, on using digress, the narrow is changed before the client receives the message in the old narrow. So, in the final state, there's one unread message caused for the user who digressed like so-

https://user-images.githubusercontent.com/55339528/112836011-a8c79880-90b7-11eb-9ee5-8881e60bb702.mp4

This is also the case with moving topics, which uses the same functions for having the notification bot send messages. Is this desired? Should I, in a prep commit make it so that we can send `mark_as_read` argument to `do_send_messages` from `send_message_moved_breadcrumbs`?

abhijeetbodas2001 committed:
zcommand: Allow sending extra data to backend.

* Clients will need to send extra data (apart from the
command itself) for upcoming new slash commands like
/giphy and /digress.

* These less simplistic commands will also come with
the possiblity of sending bad data, so we make provision
for showing custom error messages above the compose box.

abhijeetbodas2001 committed:
zcommand: Add backend code for digress slash command.

This commit adds backend code for the digress
slash command, which will send two tombstone
messages to old and new threads when used, so
as to suggest moving the current conversation
to another topic.

Before sending the messages, we check that-
* The old and new (stream+topic) combination is not
  the same.
* The user can themselves access as well as send
  messages to the involved streams.
* The old topic isn't empty for the user.

abhijeetbodas2001 committed:
zcommand: Add frontend code for digress slash command.

This commit adds changes in the frontend for the
new /digress slash command.

After a successful digress, the narrow will be
changed to the new topic for the user, and a
feedback widget will be shown.

Errors, if any, are shown above the compose box.
